File: bindings/python/satox_bindings/wallet_manager.py
"""
Satox SDK Wallet Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError
from .types import Wallet

class WalletManager:
    """Wrapper for the C++ WalletManager"""
    
    def __init__(self):
        self._core = None
        self._initialized = False
        self._callbacks = []
    
    async def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize wallet manager"""
        if self._initialized:
            raise SatoxError("WalletManager already initialized")
        
        try:
            result = await self._call_native("initialize", config)
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize WalletManager: {str(e)}")
    
    async def create_wallet(self, password: Optional[str] = None) -> Wallet:
        """Create a new wallet"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("create_wallet", password)
            return Wallet(**result)
        except Exception as e:
            raise SatoxError(f"Failed to create wallet: {str(e)}")
    
    async def load_wallet(self, wallet_path: str, password: Optional[str] = None) -> Wallet:
        """Load existing wallet"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("load_wallet", wallet_path, password)
            return Wallet(**result)
        except Exception as e:
            raise SatoxError(f"Failed to load wallet: {str(e)}")
    
    async def get_balance(self, address: str) -> Dict[str, float]:
        """Get wallet balance"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("get_balance", address)
        except Exception as e:
            raise SatoxError(f"Failed to get balance: {str(e)}")
    
    async def send_transaction(self, from_address: str, to_address: str, amount: float, asset_id: Optional[str] = None) -> str:
        """Send transaction"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("send_transaction", from_address, to_address, amount, asset_id)
        except Exception as e:
            raise SatoxError(f"Failed to send transaction: {str(e)}")
    
    async def get_transaction_history(self, address: str) -> List[Dict[str, Any]]:
        """Get transaction history"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("get_transaction_history", address)
        except Exception as e:
            raise SatoxError(f"Failed to get transaction history: {str(e)}")
    
    def _ensure_initialized(self):
        if not self._initialized:
            raise SatoxError("WalletManager not initialized")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        raise NotImplementedError("Native binding not implemented") 

File: bindings/typescript/src/wallet-manager.ts
/**
 * Satox SDK Wallet Manager Binding
 */

import { SatoxError } from './error';
import { Wallet } from './types';

export class WalletManager {
    private core: any;
    private initialized: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('WalletManager already initialized');
        }

        try {
            const result = await this.core.walletInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize WalletManager: ${error.message}`);
        }
    }

    async createWallet(password?: string): Promise<Wallet> {
        this.ensureInitialized();

        try {
            return await this.core.walletCreate(password);
        } catch (error) {
            throw new SatoxError(`Failed to create wallet: ${error.message}`);
        }
    }

    async loadWallet(walletPath: string, password?: string): Promise<Wallet> {
        this.ensureInitialized();

        try {
            return await this.core.walletLoad(walletPath, password);
        } catch (error) {
            throw new SatoxError(`Failed to load wallet: ${error.message}`);
        }
    }

    async getBalance(address: string): Promise<Record<string, number>> {
        this.ensureInitialized();

        try {
            return await this.core.walletGetBalance(address);
        } catch (error) {
            throw new SatoxError(`Failed to get balance: ${error.message}`);
        }
    }

    async sendTransaction(fromAddress: string, toAddress: string, amount: number, assetId?: string): Promise<string> {
        this.ensureInitialized();

        try {
            return await this.core.walletSendTransaction(fromAddress, toAddress, amount, assetId);
        } catch (error) {
            throw new SatoxError(`Failed to send transaction: ${error.message}`);
        }
    }

    async getTransactionHistory(address: string): Promise<Record<string, any>[]> {
        this.ensureInitialized();

        try {
            return await this.core.walletGetTransactionHistory(address);
        } catch (error) {
            throw new SatoxError(`Failed to get transaction history: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('WalletManager not initialized');
        }
    }
} 